// AUTO-GENERATED by atrius-fhir-generator (R4 MolecularSequence)
use crate::r4::*;

/// FHIR MolecularSequence type
/// 
/// Raw data describing a biological sequence.
/// 
/// ## Type: Resource type
/// Base type: http://hl7.org/fhir/StructureDefinition/DomainResource
/// 
/// ## Status: draft
/// FHIR Version: 4.0.1
/// 
/// See: [MolecularSequence](http://hl7.org/fhir/StructureDefinition/MolecularSequence)
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="dom-2", severity="error", human="If the resource is contained in another resource, it SHALL NOT contain nested Resources", expr="contained.contained.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-3", severity="error", human="If the resource is contained in another resource, it SHALL be referred to from elsewhere in the resource or SHALL refer to the containing resource", expr="contained.where((('#'+id in (%resource.descendants().reference | %resource.descendants().as(canonical) | %resource.descendants().as(uri) | %resource.descendants().as(url))) or descendants().where(reference = '#').exists() or descendants().where(as(canonical) = '#').exists() or descendants().where(as(canonical) = '#').exists()).not()).trace('unmatched', id).empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-4", severity="error", human="If a resource is contained in another resource, it SHALL NOT have a meta.versionId or a meta.lastUpdated", expr="contained.meta.versionId.empty() and contained.meta.lastUpdated.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-5", severity="error", human="If a resource is contained in another resource, it SHALL NOT have a security label", expr="contained.meta.security.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-6", severity="warning", human="A resource should have narrative for robust management", expr="text.`div`.exists()", path="MolecularSequence")]
#[fhir_invariant(key="msq-3", severity="error", human="Only 0 and 1 are valid for coordinateSystem", expr="coordinateSystem = 1 or coordinateSystem = 0", path="MolecularSequence")]
pub struct MolecularSequence {
    /// Logical id of this artifact
    /// 
    /// The logical id of the resource, as used in the URL for the resource. Once
    /// assigned, this value never changes.
    /// 
    /// ## Implementation Notes
    /// The only time that a resource does not have an id is when it is being
    /// submitted to the server using a create operation.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    pub id: Option<String>,
    /// Metadata about the resource
    /// 
    /// The metadata about the resource. This is content that is maintained by the
    /// infrastructure. Changes to the content might not always be associated with
    /// version changes to the resource.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.meta")]
    pub meta: Option<Meta>,
    /// A set of rules under which this content was created
    /// 
    /// A reference to a set of rules that were followed when the resource was
    /// constructed, and which must be understood when processing the content. Often,
    /// this is a reference to an implementation guide that defines the special rules
    /// along with other profiles etc.
    /// 
    /// ## Implementation Notes
    /// Asserting this rule set restricts the content to be only understood by a
    /// limited set of trading partners. This inherently limits the usefulness of the
    /// data in the long term. However, the existing health eco-system is highly
    /// fractured, and not yet ready to define, collect, and exchange data in a
    /// generally computable sense. Wherever possible, implementers and/or
    /// specification writers should avoid using this element. Often, when used, the
    /// URL is a reference to an implementation guide that defines these special
    /// rules as part of it's narrative along with other profiles, value sets, etc.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Modifier element - This element is labeled as a modifier because the implicit rules may provide additional knowledge about the resource that modifies it's meaning or interpretation
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.implicitRules")]
    #[fhir_serde(rename = "implicitRules")]
    pub implicit_rules: Option<Uri>,
    /// Language of the resource content
    /// 
    /// The base language in which the resource is written.
    /// 
    /// ## Implementation Notes
    /// Language is provided to support indexing and accessibility (typically,
    /// services such as text to speech use the language tag). The html language tag
    /// in the narrative applies to the narrative. The language tag on the resource
    /// may be used to specify the language of other presentations generated from the
    /// data in the resource. Not all the content has to be in the base language. The
    /// Resource.language should not be assumed to apply to the narrative
    /// automatically. If a language is specified, it should it also be specified on
    /// the div element in the html (see rules in HTML5 for information about the
    /// relationship between xml:lang and the html lang attribute).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: preferred
    /// - **Description**: A human language.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/languages
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.language")]
    pub language: Option<Code>,
    /// Text summary of the resource, for human interpretation
    /// 
    /// A human-readable narrative that contains a summary of the resource and can be
    /// used to represent the content of the resource to a human. The narrative need
    /// not encode all the structured data, but is required to contain sufficient
    /// detail to make it "clinically safe" for a human to just read the narrative.
    /// Resource definitions may define what content should be represented in the
    /// narrative to ensure clinical safety.
    /// 
    /// ## Implementation Notes
    /// Contained resources do not have narrative. Resources that are not contained
    /// SHOULD have a narrative. In some cases, a resource may only have text with
    /// little or no additional discrete data (as long as all minOccurs=1 elements
    /// are satisfied). This may be necessary for data from legacy systems where
    /// information is captured as a "text blob" or where text is additionally
    /// entered raw or narrated and encoded information is added later.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Aliases
    /// narrative, html, xhtml, display
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.text")]
    pub text: Option<Narrative>,
    /// Contained, inline Resources
    /// 
    /// These resources do not have an independent existence apart from the resource
    /// that contains them - they cannot be identified independently, and nor can
    /// they have their own independent transaction scope.
    /// 
    /// ## Implementation Notes
    /// This should never be done when the content can be identified properly, as
    /// once identification is lost, it is extremely difficult (and context
    /// dependent) to restore it again. Contained resources may have profiles and
    /// tags In their meta elements, but SHALL NOT have security labels.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Aliases
    /// inline resources, anonymous resources, contained resources
    pub contained: Option<Vec<Resource>>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the resource. To make the use of extensions safe and
    /// manageable, there is a strict set of governance applied to the definition and
    /// use of extensions. Though any implementer can define an extension, there is a
    /// set of requirements that SHALL be met as part of the definition of the
    /// extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the resource and that modifies the understanding of the element
    /// that contains it and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer is allowed to define an extension, there is a set of requirements
    /// that SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the resource that contains them
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Unique ID for this particular sequence. This is a FHIR-defined id
    /// 
    /// A unique identifier for this particular sequence instance. This is a
    /// FHIR-defined id.
    /// 
    /// ## Requirements
    /// Allows sequences to be distinguished and referenced.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.identifier")]
    pub identifier: Option<Vec<Identifier>>,
    /// aa | dna | rna
    /// 
    /// Amino Acid Sequence/ DNA Sequence / RNA Sequence.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type if a sequence -- DNA, RNA, or amino acid sequence.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/sequence-type|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.type")]
    #[fhir_serde(rename = "type")]
    pub r#type: Option<Code>,
    /// Base number of coordinate system (0 for 0-based numbering or coordinates, inclusive start, exclusive end, 1 for 1-based numbering, inclusive start, inclusive end)
    /// 
    /// Whether the sequence is numbered starting at 0 (0-based numbering or
    /// coordinates, inclusive start, exclusive end) or starting at 1 (1-based
    /// numbering, inclusive start and inclusive end).
    /// 
    /// ## Cardinality: Required (1..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.coordinateSystem")]
    #[fhir_serde(rename = "coordinateSystem")]
    pub coordinate_system: Integer,
    /// Who and/or what this is about
    /// 
    /// The patient whose sequencing results are described by this resource.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.patient")]
    pub patient: Option<Reference>,
    /// Specimen used for sequencing
    /// 
    /// Specimen used for sequencing.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.specimen")]
    pub specimen: Option<Reference>,
    /// The method for sequencing
    /// 
    /// The method for sequencing, for example, chip information.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.device")]
    pub device: Option<Reference>,
    /// Who should be responsible for test result
    /// 
    /// The organization or lab that should be responsible for this result.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.performer")]
    pub performer: Option<Reference>,
    /// The number of copies of the sequence of interest.  (RNASeq)
    /// 
    /// The number of copies of the sequence of interest. (RNASeq).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quantity")]
    pub quantity: Option<Quantity>,
    /// A sequence used as reference
    /// 
    /// A sequence that is used as a reference to describe variants that are present
    /// in a sequence analyzed.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **msq-5**: GenomeBuild and chromosome must be both contained if either one of them is contained (error)
    ///   Expression: `(chromosome.empty() and genomeBuild.empty()) or (chromosome.exists() and genomeBuild.exists())`
    /// - **msq-6**: Have and only have one of the following elements in referenceSeq : 1. genomeBuild ; 2 referenceSeqId; 3. referenceSeqPointer;  4. referenceSeqString; (error)
    ///   Expression: `(genomeBuild.count()+referenceSeqId.count()+ referenceSeqPointer.count()+ referenceSeqString.count()) = 1`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq")]
    #[fhir_invariant(key="msq-5", severity="error", human="GenomeBuild and chromosome must be both contained if either one of them is contained", expr="(chromosome.empty() and genomeBuild.empty()) or (chromosome.exists() and genomeBuild.exists())", path="MolecularSequence.referenceSeq")]
    #[fhir_invariant(key="msq-6", severity="error", human="Have and only have one of the following elements in referenceSeq : 1. genomeBuild ; 2 referenceSeqId; 3. referenceSeqPointer;  4. referenceSeqString;", expr="(genomeBuild.count()+referenceSeqId.count()+ referenceSeqPointer.count()+ referenceSeqString.count()) = 1", path="MolecularSequence.referenceSeq")]
    #[fhir_serde(rename = "referenceSeq")]
    pub reference_seq: Option<MolecularSequenceReferenceSeq>,
    /// Variant in sequence
    /// 
    /// The definition of variant here originates from Sequence ontology
    /// ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)).
    /// This element can represent amino acid or nucleic sequence change(including
    /// insertion,deletion,SNP,etc.) It can represent some complex mutation or
    /// segment variation with the assist of CIGAR string.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant")]
    pub variant: Option<Vec<MolecularSequenceVariant>>,
    /// Sequence that was observed
    /// 
    /// Sequence that was observed. It is the result marked by referenceSeq along
    /// with variant records on referenceSeq. This shall start from
    /// referenceSeq.windowStart and end by referenceSeq.windowEnd.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.observedSeq")]
    #[fhir_serde(rename = "observedSeq")]
    pub observed_seq: Option<String>,
    /// An set of value as quality of sequence
    /// 
    /// An experimental feature attribute that defines the quality of the feature in
    /// a quantitative way, such as a phred quality score
    /// ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality")]
    pub quality: Option<Vec<MolecularSequenceQuality>>,
    /// Average number of reads representing a given nucleotide in the reconstructed sequence
    /// 
    /// Coverage (read depth or depth) is the average number of reads representing a
    /// given nucleotide in the reconstructed sequence.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.readCoverage")]
    #[fhir_serde(rename = "readCoverage")]
    pub read_coverage: Option<Integer>,
    /// External repository which contains detailed report related with observedSeq in this resource
    /// 
    /// Configurations of the external repository. The repository shall store
    /// target's observedSeq or records related with target's observedSeq.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository")]
    pub repository: Option<Vec<MolecularSequenceRepository>>,
    /// Pointer to next atomic sequence
    /// 
    /// Pointer to next atomic sequence which at most contains one variant.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.pointer")]
    pub pointer: Option<Vec<Reference>>,
    /// Structural variant
    /// 
    /// Information about chromosome structure variation.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant")]
    #[fhir_serde(rename = "structureVariant")]
    pub structure_variant: Option<Vec<MolecularSequenceStructureVariant>>,
}

/// An set of value as quality of sequence
/// 
/// An experimental feature attribute that defines the quality of the feature in
/// a quantitative way, such as a phred quality score
/// ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality")]
pub struct MolecularSequenceQuality {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.quality.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.quality.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// indel | snp | unknown
    /// 
    /// INDEL / SNP / Undefined variant.
    /// 
    /// ## Cardinality: Required (1..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for quality report.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/quality-type|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.type")]
    #[fhir_serde(rename = "type")]
    pub r#type: Code,
    /// Standard sequence for comparison
    /// 
    /// Gold standard sequence used for comparing against.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: example
    /// - **Description**: Reference identifier of the sequence that used to mark the quality of tested samples.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/sequence-quality-standardSequence
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.standardSequence")]
    #[fhir_serde(rename = "standardSequence")]
    pub standard_sequence: Option<CodeableConcept>,
    /// Start position of the sequence
    /// 
    /// Start position of the sequence. If the coordinate system is either 0-based or
    /// 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.start")]
    pub start: Option<Integer>,
    /// End position of the sequence
    /// 
    /// End position of the sequence. If the coordinate system is 0-based then end is
    /// exclusive and does not include the last position. If the coordinate system is
    /// 1-base, then end is inclusive and includes the last position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.end")]
    pub end: Option<Integer>,
    /// Quality score for the comparison
    /// 
    /// The score of an experimentally derived feature such as a p-value
    /// ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.score")]
    pub score: Option<Quantity>,
    /// Method to get quality
    /// 
    /// Which method is used to get sequence quality.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: example
    /// - **Description**: The method used to evaluate the numerical quality of the observed sequence.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/sequence-quality-method
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.method")]
    pub method: Option<CodeableConcept>,
    /// True positives from the perspective of the truth data
    /// 
    /// True positives, from the perspective of the truth data, i.e. the number of
    /// sites in the Truth Call Set for which there are paths through the Query Call
    /// Set that are consistent with all of the alleles at this site, and for which
    /// there is an accurate genotype call for the event.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.truthTP")]
    #[fhir_serde(rename = "truthTP")]
    pub truth_t_p: Option<Decimal>,
    /// True positives from the perspective of the query data
    /// 
    /// True positives, from the perspective of the query data, i.e. the number of
    /// sites in the Query Call Set for which there are paths through the Truth Call
    /// Set that are consistent with all of the alleles at this site, and for which
    /// there is an accurate genotype call for the event.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.queryTP")]
    #[fhir_serde(rename = "queryTP")]
    pub query_t_p: Option<Decimal>,
    /// False negatives
    /// 
    /// False negatives, i.e. the number of sites in the Truth Call Set for which
    /// there is no path through the Query Call Set that is consistent with all of
    /// the alleles at this site, or sites for which there is an inaccurate genotype
    /// call for the event. Sites with correct variant but incorrect genotype are
    /// counted here.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.truthFN")]
    #[fhir_serde(rename = "truthFN")]
    pub truth_f_n: Option<Decimal>,
    /// False positives
    /// 
    /// False positives, i.e. the number of sites in the Query Call Set for which
    /// there is no path through the Truth Call Set that is consistent with this
    /// site. Sites with correct variant but incorrect genotype are counted here.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.queryFP")]
    #[fhir_serde(rename = "queryFP")]
    pub query_f_p: Option<Decimal>,
    /// False positives where the non-REF alleles in the Truth and Query Call Sets match
    /// 
    /// The number of false positives where the non-REF alleles in the Truth and
    /// Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1
    /// or similar).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.gtFP")]
    #[fhir_serde(rename = "gtFP")]
    pub gt_f_p: Option<Decimal>,
    /// Precision of comparison
    /// 
    /// QUERY.TP / (QUERY.TP + QUERY.FP).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.precision")]
    pub precision: Option<Decimal>,
    /// Recall of comparison
    /// 
    /// TRUTH.TP / (TRUTH.TP + TRUTH.FN).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.recall")]
    pub recall: Option<Decimal>,
    /// F-score
    /// 
    /// Harmonic mean of Recall and Precision, computed as: 2 * precision * recall /
    /// (precision + recall).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.fScore")]
    #[fhir_serde(rename = "fScore")]
    pub f_score: Option<Decimal>,
    /// Receiver Operator Characteristic (ROC) Curve
    /// 
    /// Receiver Operator Characteristic (ROC) Curve to give sensitivity/specificity
    /// tradeoff.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc")]
    pub roc: Option<MolecularSequenceQualityRoc>,
}

/// Receiver Operator Characteristic (ROC) Curve
/// 
/// Receiver Operator Characteristic (ROC) Curve to give sensitivity/specificity
/// tradeoff.
/// 
/// ## Cardinality: Optional (0..1)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc")]
pub struct MolecularSequenceQualityRoc {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.quality.roc.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.quality.roc.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Genotype quality score
    /// 
    /// Invidual data point representing the GQ (genotype quality) score threshold.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.score")]
    pub score: Option<Vec<Integer>>,
    /// Roc score true positive numbers
    /// 
    /// The number of true positives if the GQ score threshold was set to "score"
    /// field value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.numTP")]
    #[fhir_serde(rename = "numTP")]
    pub num_t_p: Option<Vec<Integer>>,
    /// Roc score false positive numbers
    /// 
    /// The number of false positives if the GQ score threshold was set to "score"
    /// field value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.numFP")]
    #[fhir_serde(rename = "numFP")]
    pub num_f_p: Option<Vec<Integer>>,
    /// Roc score false negative numbers
    /// 
    /// The number of false negatives if the GQ score threshold was set to "score"
    /// field value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.numFN")]
    #[fhir_serde(rename = "numFN")]
    pub num_f_n: Option<Vec<Integer>>,
    /// Precision of the GQ score
    /// 
    /// Calculated precision if the GQ score threshold was set to "score" field
    /// value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.precision")]
    pub precision: Option<Vec<Decimal>>,
    /// Sensitivity of the GQ score
    /// 
    /// Calculated sensitivity if the GQ score threshold was set to "score" field
    /// value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.sensitivity")]
    pub sensitivity: Option<Vec<Decimal>>,
    /// FScore of the GQ score
    /// 
    /// Calculated fScore if the GQ score threshold was set to "score" field value.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.quality.roc.fMeasure")]
    #[fhir_serde(rename = "fMeasure")]
    pub f_measure: Option<Vec<Decimal>>,
}

/// A sequence used as reference
/// 
/// A sequence that is used as a reference to describe variants that are present
/// in a sequence analyzed.
/// 
/// ## Cardinality: Optional (0..1)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
/// - **msq-5**: GenomeBuild and chromosome must be both contained if either one of them is contained (error)
///   Expression: `(chromosome.empty() and genomeBuild.empty()) or (chromosome.exists() and genomeBuild.exists())`
/// - **msq-6**: Have and only have one of the following elements in referenceSeq : 1. genomeBuild ; 2 referenceSeqId; 3. referenceSeqPointer;  4. referenceSeqString; (error)
///   Expression: `(genomeBuild.count()+referenceSeqId.count()+ referenceSeqPointer.count()+ referenceSeqString.count()) = 1`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq")]
#[fhir_invariant(key="msq-5", severity="error", human="GenomeBuild and chromosome must be both contained if either one of them is contained", expr="(chromosome.empty() and genomeBuild.empty()) or (chromosome.exists() and genomeBuild.exists())", path="MolecularSequence.referenceSeq")]
#[fhir_invariant(key="msq-6", severity="error", human="Have and only have one of the following elements in referenceSeq : 1. genomeBuild ; 2 referenceSeqId; 3. referenceSeqPointer;  4. referenceSeqString;", expr="(genomeBuild.count()+referenceSeqId.count()+ referenceSeqPointer.count()+ referenceSeqString.count()) = 1", path="MolecularSequence.referenceSeq")]
pub struct MolecularSequenceReferenceSeq {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.referenceSeq.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.referenceSeq.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Chromosome containing genetic finding
    /// 
    /// Structural unit composed of a nucleic acid molecule which controls its own
    /// replication through the interaction of specific proteins at one or more
    /// origins of replication
    /// ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: example
    /// - **Description**: Chromosome number for human.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/chromosome-human
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.chromosome")]
    pub chromosome: Option<CodeableConcept>,
    /// The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'
    /// 
    /// The Genome Build used for reference, following GRCh build versions e.g. 'GRCh
    /// 37'. Version number must be included if a versioned release of a primary
    /// build was used.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.genomeBuild")]
    #[fhir_serde(rename = "genomeBuild")]
    pub genome_build: Option<String>,
    /// sense | antisense
    /// 
    /// A relative reference to a DNA strand based on gene orientation. The strand
    /// that contains the open reading frame of the gene is the "sense" strand, and
    /// the opposite complementary strand is the "antisense" strand.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for orientation.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/orientation-type|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.orientation")]
    pub orientation: Option<Code>,
    /// Reference identifier
    /// 
    /// Reference identifier of reference sequence submitted to NCBI. It must match
    /// the type in the MolecularSequence.type field. For example, the prefix,
    /// NG_ identifies reference sequence for genes, NM_ for messenger
    /// RNA transcripts, and NP_ for amino acid sequences.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: example
    /// - **Description**: Reference identifier.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/sequence-referenceSeq
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.referenceSeqId")]
    #[fhir_serde(rename = "referenceSeqId")]
    pub reference_seq_id: Option<CodeableConcept>,
    /// A pointer to another MolecularSequence entity as reference sequence
    /// 
    /// A pointer to another MolecularSequence entity as reference sequence.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.referenceSeqPointer")]
    #[fhir_serde(rename = "referenceSeqPointer")]
    pub reference_seq_pointer: Option<Reference>,
    /// A string to represent reference sequence
    /// 
    /// A string like "ACGT".
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.referenceSeqString")]
    #[fhir_serde(rename = "referenceSeqString")]
    pub reference_seq_string: Option<String>,
    /// watson | crick
    /// 
    /// An absolute reference to a strand. The Watson strand is the strand whose
    /// 5'-end is on the short arm of the chromosome, and the Crick strand as the one
    /// whose 5'-end is on the long arm.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for strand.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/strand-type|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.strand")]
    pub strand: Option<Code>,
    /// Start position of the window on the  reference sequence
    /// 
    /// Start position of the window on the reference sequence. If the coordinate
    /// system is either 0-based or 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.windowStart")]
    #[fhir_serde(rename = "windowStart")]
    pub window_start: Option<Integer>,
    /// End position of the window on the reference sequence
    /// 
    /// End position of the window on the reference sequence. If the coordinate
    /// system is 0-based then end is exclusive and does not include the last
    /// position. If the coordinate system is 1-base, then end is inclusive and
    /// includes the last position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.referenceSeq.windowEnd")]
    #[fhir_serde(rename = "windowEnd")]
    pub window_end: Option<Integer>,
}

/// External repository which contains detailed report related with observedSeq in this resource
/// 
/// Configurations of the external repository. The repository shall store
/// target's observedSeq or records related with target's observedSeq.
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository")]
pub struct MolecularSequenceRepository {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.repository.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.repository.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// directlink | openapi | login | oauth | other
    /// 
    /// Click and see / RESTful API / Need login to see / RESTful API with
    /// authentication / Other ways to see resource.
    /// 
    /// ## Cardinality: Required (1..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for access of external URI.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/repository-type|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.type")]
    #[fhir_serde(rename = "type")]
    pub r#type: Code,
    /// URI of the repository
    /// 
    /// URI of an external repository which contains further details about the
    /// genetics data.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.url")]
    pub url: Option<Uri>,
    /// Repository's name
    /// 
    /// URI of an external repository which contains further details about the
    /// genetics data.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.name")]
    pub name: Option<String>,
    /// Id of the dataset that used to call for dataset in repository
    /// 
    /// Id of the variant in this external repository. The server will understand how
    /// to use this id to call for more info about datasets in external repository.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.datasetId")]
    #[fhir_serde(rename = "datasetId")]
    pub dataset_id: Option<String>,
    /// Id of the variantset that used to call for variantset in repository
    /// 
    /// Id of the variantset in this external repository. The server will understand
    /// how to use this id to call for more info about variantsets in external
    /// repository.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.variantsetId")]
    #[fhir_serde(rename = "variantsetId")]
    pub variantset_id: Option<String>,
    /// Id of the read
    /// 
    /// Id of the read in this external repository.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.repository.readsetId")]
    #[fhir_serde(rename = "readsetId")]
    pub readset_id: Option<String>,
}

/// Structural variant
/// 
/// Information about chromosome structure variation.
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant")]
pub struct MolecularSequenceStructureVariant {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Structural variant change type
    /// 
    /// Information about chromosome structure variation DNA change type.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: DNA change type.
    /// - **ValueSet**: http://loinc.org/vs/LL379-9|4.0.1
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.variantType")]
    #[fhir_serde(rename = "variantType")]
    pub variant_type: Option<CodeableConcept>,
    /// Does the structural variant have base pair resolution breakpoints?
    /// 
    /// Used to indicate if the outer and inner start-end values have the same
    /// meaning.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.exact")]
    pub exact: Option<Boolean>,
    /// Structural variant length
    /// 
    /// Length of the variant chromosome.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.length")]
    pub length: Option<Integer>,
    /// Structural variant outer
    /// 
    /// Structural variant outer.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer")]
    pub outer: Option<MolecularSequenceStructureVariantOuter>,
    /// Structural variant inner
    /// 
    /// Structural variant inner.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner")]
    pub inner: Option<MolecularSequenceStructureVariantInner>,
}

/// Structural variant inner
/// 
/// Structural variant inner.
/// 
/// ## Cardinality: Optional (0..1)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner")]
pub struct MolecularSequenceStructureVariantInner {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.inner.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.inner.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Structural variant inner start
    /// 
    /// Structural variant inner start. If the coordinate system is either 0-based or
    /// 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner.start")]
    pub start: Option<Integer>,
    /// Structural variant inner end
    /// 
    /// Structural variant inner end. If the coordinate system is 0-based then end is
    /// exclusive and does not include the last position. If the coordinate system is
    /// 1-base, then end is inclusive and includes the last position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.inner.end")]
    pub end: Option<Integer>,
}

/// Structural variant outer
/// 
/// Structural variant outer.
/// 
/// ## Cardinality: Optional (0..1)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer")]
pub struct MolecularSequenceStructureVariantOuter {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.outer.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.structureVariant.outer.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Structural variant outer start
    /// 
    /// Structural variant outer start. If the coordinate system is either 0-based or
    /// 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer.start")]
    pub start: Option<Integer>,
    /// Structural variant outer end
    /// 
    /// Structural variant outer end. If the coordinate system is 0-based then end is
    /// exclusive and does not include the last position. If the coordinate system is
    /// 1-base, then end is inclusive and includes the last position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.structureVariant.outer.end")]
    pub end: Option<Integer>,
}

/// Variant in sequence
/// 
/// The definition of variant here originates from Sequence ontology
/// ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)).
/// This element can represent amino acid or nucleic sequence change(including
/// insertion,deletion,SNP,etc.) It can represent some complex mutation or
/// segment variation with the assist of CIGAR string.
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant")]
pub struct MolecularSequenceVariant {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and manageable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.variant.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and manageable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.variant.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Start position of the variant on the  reference sequence
    /// 
    /// Start position of the variant on the reference sequence. If the coordinate
    /// system is either 0-based or 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.start")]
    pub start: Option<Integer>,
    /// End position of the variant on the reference sequence
    /// 
    /// End position of the variant on the reference sequence. If the coordinate
    /// system is 0-based then end is exclusive and does not include the last
    /// position. If the coordinate system is 1-base, then end is inclusive and
    /// includes the last position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.end")]
    pub end: Option<Integer>,
    /// Allele that was observed
    /// 
    /// An allele is one of a set of coexisting sequence variants of a gene
    /// ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).
    /// Nucleotide(s)/amino acids from start position of sequence to stop position of
    /// sequence on the positive (+) strand of the observed sequence. When the
    /// sequence type is DNA, it should be the sequence on the positive (+) strand.
    /// This will lay in the range between variant.start and variant.end.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.observedAllele")]
    #[fhir_serde(rename = "observedAllele")]
    pub observed_allele: Option<String>,
    /// Allele in the reference sequence
    /// 
    /// An allele is one of a set of coexisting sequence variants of a gene
    /// ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).
    /// Nucleotide(s)/amino acids from start position of sequence to stop position of
    /// sequence on the positive (+) strand of the reference sequence. When the
    /// sequence type is DNA, it should be the sequence on the positive (+) strand.
    /// This will lay in the range between variant.start and variant.end.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.referenceAllele")]
    #[fhir_serde(rename = "referenceAllele")]
    pub reference_allele: Option<String>,
    /// Extended CIGAR string for aligning the sequence with reference bases
    /// 
    /// Extended CIGAR string for aligning the sequence with reference bases. See
    /// detailed documentation
    /// [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.cigar")]
    pub cigar: Option<String>,
    /// Pointer to observed variant information
    /// 
    /// A pointer to an Observation containing variant information.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.variant.variantPointer")]
    #[fhir_serde(rename = "variantPointer")]
    pub variant_pointer: Option<Reference>,
}

