// AUTO-GENERATED by atrius-fhir-generator (R5 MolecularSequence)
use crate::r5::*;

/// FHIR MolecularSequence type
/// 
/// Representation of a molecular sequence.
/// 
/// ## Type: Resource type
/// Base type: http://hl7.org/fhir/StructureDefinition/DomainResource
/// 
/// ## Status: draft
/// FHIR Version: 5.0.0
/// 
/// See: [MolecularSequence](http://hl7.org/fhir/StructureDefinition/MolecularSequence)
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="dom-2", severity="error", human="If the resource is contained in another resource, it SHALL NOT contain nested Resources", expr="contained.contained.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-3", severity="error", human="If the resource is contained in another resource, it SHALL be referred to from elsewhere in the resource or SHALL refer to the containing resource", expr="contained.where((('#'+id in (%resource.descendants().reference | %resource.descendants().ofType(canonical) | %resource.descendants().ofType(uri) | %resource.descendants().ofType(url))) or descendants().where(reference = '#').exists() or descendants().where(ofType(canonical) = '#').exists() or descendants().where(ofType(canonical) = '#').exists()).not()).trace('unmatched', id).empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-4", severity="error", human="If a resource is contained in another resource, it SHALL NOT have a meta.versionId or a meta.lastUpdated", expr="contained.meta.versionId.empty() and contained.meta.lastUpdated.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-5", severity="error", human="If a resource is contained in another resource, it SHALL NOT have a security label", expr="contained.meta.security.empty()", path="MolecularSequence")]
#[fhir_invariant(key="dom-6", severity="warning", human="A resource should have narrative for robust management", expr="text.`div`.exists()", path="MolecularSequence")]
pub struct MolecularSequence {
    /// Logical id of this artifact
    /// 
    /// The logical id of the resource, as used in the URL for the resource. Once
    /// assigned, this value never changes.
    /// 
    /// ## Implementation Notes
    /// Within the context of the FHIR RESTful interactions, the resource has an id
    /// except for cases like the create and conditional update. Otherwise, the use
    /// of the resouce id depends on the given use case.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    pub id: Option<String>,
    /// Metadata about the resource
    /// 
    /// The metadata about the resource. This is content that is maintained by the
    /// infrastructure. Changes to the content might not always be associated with
    /// version changes to the resource.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.meta")]
    pub meta: Option<Meta>,
    /// A set of rules under which this content was created
    /// 
    /// A reference to a set of rules that were followed when the resource was
    /// constructed, and which must be understood when processing the content. Often,
    /// this is a reference to an implementation guide that defines the special rules
    /// along with other profiles etc.
    /// 
    /// ## Implementation Notes
    /// Asserting this rule set restricts the content to be only understood by a
    /// limited set of trading partners. This inherently limits the usefulness of the
    /// data in the long term. However, the existing health eco-system is highly
    /// fractured, and not yet ready to define, collect, and exchange data in a
    /// generally computable sense. Wherever possible, implementers and/or
    /// specification writers should avoid using this element. Often, when used, the
    /// URL is a reference to an implementation guide that defines these special
    /// rules as part of its narrative along with other profiles, value sets, etc.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Modifier element - This element is labeled as a modifier because the implicit rules may provide additional knowledge about the resource that modifies its meaning or interpretation
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.implicitRules")]
    #[fhir_serde(rename = "implicitRules")]
    pub implicit_rules: Option<Uri>,
    /// Language of the resource content
    /// 
    /// The base language in which the resource is written.
    /// 
    /// ## Implementation Notes
    /// Language is provided to support indexing and accessibility (typically,
    /// services such as text to speech use the language tag). The html language tag
    /// in the narrative applies to the narrative. The language tag on the resource
    /// may be used to specify the language of other presentations generated from the
    /// data in the resource. Not all the content has to be in the base language. The
    /// Resource.language should not be assumed to apply to the narrative
    /// automatically. If a language is specified, it should it also be specified on
    /// the div element in the html (see rules in HTML5 for information about the
    /// relationship between xml:lang and the html lang attribute).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: IETF language tag for a human language
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/all-languages|5.0.0
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.language")]
    pub language: Option<Code>,
    /// Text summary of the resource, for human interpretation
    /// 
    /// A human-readable narrative that contains a summary of the resource and can be
    /// used to represent the content of the resource to a human. The narrative need
    /// not encode all the structured data, but is required to contain sufficient
    /// detail to make it "clinically safe" for a human to just read the narrative.
    /// Resource definitions may define what content should be represented in the
    /// narrative to ensure clinical safety.
    /// 
    /// ## Implementation Notes
    /// Contained resources do not have a narrative. Resources that are not contained
    /// SHOULD have a narrative. In some cases, a resource may only have text with
    /// little or no additional discrete data (as long as all minOccurs=1 elements
    /// are satisfied). This may be necessary for data from legacy systems where
    /// information is captured as a "text blob" or where text is additionally
    /// entered raw or narrated and encoded information is added later.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Aliases
    /// narrative, html, xhtml, display
    /// 
    /// ## Conditions
    /// Used when: dom-6
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.text")]
    pub text: Option<Narrative>,
    /// Contained, inline Resources
    /// 
    /// These resources do not have an independent existence apart from the resource
    /// that contains them - they cannot be identified independently, nor can they
    /// have their own independent transaction scope. This is allowed to be a
    /// Parameters resource if and only if it is referenced by a resource that
    /// provides context/meaning.
    /// 
    /// ## Implementation Notes
    /// This should never be done when the content can be identified properly, as
    /// once identification is lost, it is extremely difficult (and context
    /// dependent) to restore it again. Contained resources may have profiles and
    /// tags in their meta elements, but SHALL NOT have security labels.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Aliases
    /// inline resources, anonymous resources, contained resources
    /// 
    /// ## Conditions
    /// Used when: dom-2, dom-4, dom-3, dom-5
    pub contained: Option<Vec<Resource>>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the resource. To make the use of extensions safe and managable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the resource and that modifies the understanding of the element
    /// that contains it and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and managable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer is allowed to define an extension, there is a set of requirements
    /// that SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the resource that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Unique ID for this particular sequence
    /// 
    /// A unique identifier for this particular sequence instance.
    /// 
    /// ## Requirements
    /// Allows sequences to be distinguished and referenced.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.identifier")]
    pub identifier: Option<Vec<Identifier>>,
    /// aa | dna | rna
    /// 
    /// Amino Acid Sequence/ DNA Sequence / RNA Sequence.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type if a sequence -- DNA, RNA, or amino acid sequence.
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/sequence-type|5.0.0
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.type")]
    #[fhir_serde(rename = "type")]
    pub r#type: Option<Code>,
    /// Subject this sequence is associated too
    /// 
    /// Indicates the subject this sequence is associated too.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.subject")]
    pub subject: Option<Reference>,
    /// What the molecular sequence is about, when it is not about the subject of record
    /// 
    /// The actual focus of a molecular sequence when it is not the patient of record
    /// representing something or someone associated with the patient such as a
    /// spouse, parent, child, or sibling. For example, in trio testing, the subject
    /// would be the child (proband) and the focus would be the parent.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.focus")]
    pub focus: Option<Vec<Reference>>,
    /// Specimen used for sequencing
    /// 
    /// Specimen used for sequencing.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.specimen")]
    pub specimen: Option<Reference>,
    /// The method for sequencing
    /// 
    /// The method for sequencing, for example, chip information.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.device")]
    pub device: Option<Reference>,
    /// Who should be responsible for test result
    /// 
    /// The organization or lab that should be responsible for this result.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.performer")]
    pub performer: Option<Reference>,
    /// Sequence that was observed
    /// 
    /// Sequence that was observed.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.literal")]
    pub literal: Option<String>,
    /// Embedded file or a link (URL) which contains content to represent the sequence
    /// 
    /// Sequence that was observed as file content. Can be an actual file contents,
    /// or referenced by a URL to an external system.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.formatted")]
    pub formatted: Option<Vec<Attachment>>,
    /// A sequence defined relative to another sequence
    /// 
    /// A sequence defined relative to another sequence.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative")]
    pub relative: Option<Vec<MolecularSequenceRelative>>,
}

/// A sequence defined relative to another sequence
/// 
/// A sequence defined relative to another sequence.
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative")]
pub struct MolecularSequenceRelative {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Conditions
    /// Used when: ele-1
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and managable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and managable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Ways of identifying nucleotides or amino acids within a sequence
    /// 
    /// These are different ways of identifying nucleotides or amino acids within a
    /// sequence. Different databases and file types may use different systems. For
    /// detail definitions, see https://loinc.org/92822-6/ for more detail.
    /// 
    /// ## Cardinality: Required (1..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: extensible
    /// - **Description**: Genomic coordinate system.
    /// - **ValueSet**: http://loinc.org/LL5323-2/
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.coordinateSystem")]
    #[fhir_serde(rename = "coordinateSystem")]
    pub coordinate_system: CodeableConcept,
    /// Indicates the order in which the sequence should be considered when putting multiple 'relative' elements together
    /// 
    /// Indicates the order in which the sequence should be considered when putting
    /// multiple 'relative' elements together.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.ordinalPosition")]
    #[fhir_serde(rename = "ordinalPosition")]
    pub ordinal_position: Option<Integer>,
    /// Indicates the nucleotide range in the composed sequence when multiple 'relative' elements are used together
    /// 
    /// Indicates the nucleotide range in the composed sequence when multiple
    /// 'relative' elements are used together.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.sequenceRange")]
    #[fhir_serde(rename = "sequenceRange")]
    pub sequence_range: Option<Range>,
    /// A sequence used as starting sequence
    /// 
    /// A sequence that is used as a starting sequence to describe variants that are
    /// present in a sequence analyzed.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **msq-5**: Both genomeAssembly and chromosome must be both contained if either one of them is contained (error)
    ///   Expression: `chromosome.exists() = genomeAssembly.exists()`
    /// - **msq-6**: Have and only have one of the following elements in startingSequence: 1. genomeAssembly; 2 sequence (error)
    ///   Expression: `genomeAssembly.exists() xor sequence.exists()`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence")]
    #[fhir_invariant(key="msq-5", severity="error", human="Both genomeAssembly and chromosome must be both contained if either one of them is contained", expr="chromosome.exists() = genomeAssembly.exists()", path="MolecularSequence.relative.startingSequence")]
    #[fhir_invariant(key="msq-6", severity="error", human="Have and only have one of the following elements in startingSequence: 1. genomeAssembly; 2 sequence", expr="genomeAssembly.exists() xor sequence.exists()", path="MolecularSequence.relative.startingSequence")]
    #[fhir_serde(rename = "startingSequence")]
    pub starting_sequence: Option<MolecularSequenceRelativeStartingSequence>,
    /// Changes in sequence from the starting sequence
    /// 
    /// Changes in sequence from the starting sequence.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit")]
    pub edit: Option<Vec<MolecularSequenceRelativeEdit>>,
}

/// Changes in sequence from the starting sequence
/// 
/// Changes in sequence from the starting sequence.
/// 
/// ## Cardinality: Optional, Multiple (0..*)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit")]
pub struct MolecularSequenceRelativeEdit {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Conditions
    /// Used when: ele-1
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and managable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.edit.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and managable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.edit.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// Start position of the edit on the starting sequence
    /// 
    /// Start position of the edit on the starting sequence. If the coordinate system
    /// is either 0-based or 1-based, then start position is inclusive.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.start")]
    pub start: Option<Integer>,
    /// End position of the edit on the starting sequence
    /// 
    /// End position of the edit on the starting sequence. If the coordinate system
    /// is 0-based then end is exclusive and does not include the last position. If
    /// the coordinate system is 1-base, then end is inclusive and includes the last
    /// position.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.end")]
    pub end: Option<Integer>,
    /// Allele that was observed
    /// 
    /// Allele that was observed. Nucleotide(s)/amino acids from start position of
    /// sequence to stop position of sequence on the positive (+) strand of the
    /// observed sequence. When the sequence type is DNA, it should be the sequence
    /// on the positive (+) strand. This will lay in the range between variant.start
    /// and variant.end.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.replacementSequence")]
    #[fhir_serde(rename = "replacementSequence")]
    pub replacement_sequence: Option<String>,
    /// Allele in the starting sequence
    /// 
    /// Allele in the starting sequence. Nucleotide(s)/amino acids from start
    /// position of sequence to stop position of sequence on the positive (+) strand
    /// of the starting sequence. When the sequence type is DNA, it should be the
    /// sequence on the positive (+) strand. This will lay in the range between
    /// variant.start and variant.end.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.edit.replacedSequence")]
    #[fhir_serde(rename = "replacedSequence")]
    pub replaced_sequence: Option<String>,
}

/// Choice of types for the sequence\[x\] field in MolecularSequenceRelativeStartingSequence
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath)]
#[fhir_choice_element(base_name = "sequence")]
pub enum MolecularSequenceRelativeStartingSequenceSequence {
    /// Variant accepting the CodeableConcept type.
    #[fhir_serde(rename = "sequenceCodeableConcept")]
    CodeableConcept(CodeableConcept),
    /// Variant accepting the String type.
    #[fhir_serde(rename = "sequenceString")]
    String(String),
    /// Variant accepting the Reference type.
    #[fhir_serde(rename = "sequenceReference")]
    Reference(Reference),
}

/// A sequence used as starting sequence
/// 
/// A sequence that is used as a starting sequence to describe variants that are
/// present in a sequence analyzed.
/// 
/// ## Cardinality: Optional (0..1)
/// 
/// ## Special Semantics
/// - Included in summary
/// 
/// ## Constraints
/// - **ele-1**: All FHIR elements must have a @value or children (error)
///   Expression: `hasValue() or (children().count() > id.count())`
/// - **msq-5**: Both genomeAssembly and chromosome must be both contained if either one of them is contained (error)
///   Expression: `chromosome.exists() = genomeAssembly.exists()`
/// - **msq-6**: Have and only have one of the following elements in startingSequence: 1. genomeAssembly; 2 sequence (error)
///   Expression: `genomeAssembly.exists() xor sequence.exists()`
#[derive(Debug, Clone, PartialEq, FhirSerde, FhirPath, Default, FhirValidate)]
#[fhir_resource(choice_elements = "sequence")]
#[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence")]
#[fhir_invariant(key="msq-5", severity="error", human="Both genomeAssembly and chromosome must be both contained if either one of them is contained", expr="chromosome.exists() = genomeAssembly.exists()", path="MolecularSequence.relative.startingSequence")]
#[fhir_invariant(key="msq-6", severity="error", human="Have and only have one of the following elements in startingSequence: 1. genomeAssembly; 2 sequence", expr="genomeAssembly.exists() xor sequence.exists()", path="MolecularSequence.relative.startingSequence")]
pub struct MolecularSequenceRelativeStartingSequence {
    /// Unique id for inter-element referencing
    /// 
    /// Unique id for the element within a resource (for internal references). This
    /// may be any string value that does not contain spaces.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Conditions
    /// Used when: ele-1
    pub id: Option<String>,
    /// Additional content defined by implementations
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element. To make the use of extensions safe and managable,
    /// there is a strict set of governance applied to the definition and use of
    /// extensions. Though any implementer can define an extension, there is a set of
    /// requirements that SHALL be met as part of the definition of the extension.
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.extension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.startingSequence.extension")]
    pub extension: Option<Vec<Extension>>,
    /// Extensions that cannot be ignored even if unrecognized
    /// 
    /// May be used to represent additional information that is not part of the basic
    /// definition of the element and that modifies the understanding of the element
    /// in which it is contained and/or the understanding of the containing element's
    /// descendants. Usually modifier elements provide negation or qualification. To
    /// make the use of extensions safe and managable, there is a strict set of
    /// governance applied to the definition and use of extensions. Though any
    /// implementer can define an extension, there is a set of requirements that
    /// SHALL be met as part of the definition of the extension. Applications
    /// processing a resource are required to check for modifier extensions.
    /// 
    /// Modifier extensions SHALL NOT change the meaning of any elements on Resource
    /// or DomainResource (including cannot change the meaning of modifierExtension
    /// itself).
    /// 
    /// ## Requirements
    /// Modifier extensions allow for extensions that *cannot* be safely ignored to
    /// be clearly distinguished from the vast majority of extensions which can be
    /// safely ignored. This promotes interoperability by eliminating the need for
    /// implementers to prohibit the presence of extensions. For further information,
    /// see the [definition of modifier
    /// extensions](extensibility.html#modifierExtension).
    /// 
    /// ## Implementation Notes
    /// There can be no stigma associated with the use of extensions by any
    /// application, project, or standard - regardless of the institution or
    /// jurisdiction that uses or defines the extensions. The use of extensions is
    /// what allows the FHIR specification to retain a core level of simplicity for
    /// everyone.
    /// 
    /// ## Cardinality: Optional, Multiple (0..*)
    /// 
    /// ## Special Semantics
    /// - Modifier element - Modifier extensions are expected to modify the meaning or interpretation of the element that contains them
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// - **ext-1**: Must have either extensions or value[x], not both (error)
    ///   Expression: `extension.exists() != value.exists()`
    /// 
    /// ## Aliases
    /// extensions, user content, modifiers
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.modifierExtension")]
    #[fhir_invariant(key="ext-1", severity="error", human="Must have either extensions or value[x], not both", expr="extension.exists() != value.exists()", path="MolecularSequence.relative.startingSequence.modifierExtension")]
    #[fhir_serde(rename = "modifierExtension")]
    pub modifier_extension: Option<Vec<Extension>>,
    /// The genome assembly used for starting sequence, e.g. GRCh38
    /// 
    /// The genome assembly used for starting sequence, e.g. GRCh38.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: extensible
    /// - **Description**: Human reference sequence NCBI build ID.
    /// - **ValueSet**: http://loinc.org/LL1040-6/
    /// 
    /// ## Conditions
    /// Used when: msq-6, msq-5
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.genomeAssembly")]
    #[fhir_serde(rename = "genomeAssembly")]
    pub genome_assembly: Option<CodeableConcept>,
    /// Chromosome Identifier
    /// 
    /// Structural unit composed of a nucleic acid molecule which controls its own
    /// replication through the interaction of specific proteins at one or more
    /// origins of replication
    /// ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: The chromosome containing the sequence.
    /// - **ValueSet**: http://loinc.org/LL2938-0/|5.0.0
    /// 
    /// ## Conditions
    /// Used when: msq-5
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.chromosome")]
    pub chromosome: Option<CodeableConcept>,
    /// The reference sequence that represents the starting sequence
    /// 
    /// The reference sequence that represents the starting sequence.
    /// 
    /// ## Implementation Notes
    /// A starting sequence may be represented in one of three ways:
    ///   1.  CodeableConcept, using NCBI, LRG or other appropriate code systems
    ///   1.  a simple string of IUPAC codes
    ///   1.  a reference to another MolecularSequence resource.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: example
    /// - **Description**: Multiple bindings acceptable (NCBI or LRG)
    /// 
    /// ## Conditions
    /// Used when: msq-6
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.sequence[x]")]
    #[fhir_serde(flatten)]
    pub sequence: Option<MolecularSequenceRelativeStartingSequenceSequence>,
    /// Start position of the window on the starting sequence
    /// 
    /// Start position of the window on the starting sequence. This value should
    /// honor the rules of the coordinateSystem.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.windowStart")]
    #[fhir_serde(rename = "windowStart")]
    pub window_start: Option<Integer>,
    /// End position of the window on the starting sequence
    /// 
    /// End position of the window on the starting sequence. This value should honor
    /// the rules of the coordinateSystem.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.windowEnd")]
    #[fhir_serde(rename = "windowEnd")]
    pub window_end: Option<Integer>,
    /// sense | antisense
    /// 
    /// A relative reference to a DNA strand based on gene orientation. The strand
    /// that contains the open reading frame of the gene is the "sense" strand, and
    /// the opposite complementary strand is the "antisense" strand.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for orientation
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/orientation-type|5.0.0
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.orientation")]
    pub orientation: Option<Code>,
    /// watson | crick
    /// 
    /// An absolute reference to a strand. The Watson strand is the strand whose
    /// 5'-end is on the short arm of the chromosome, and the Crick strand as the one
    /// whose 5'-end is on the long arm.
    /// 
    /// ## Cardinality: Optional (0..1)
    /// 
    /// ## Special Semantics
    /// - Included in summary
    /// 
    /// ## Constraints
    /// - **ele-1**: All FHIR elements must have a @value or children (error)
    ///   Expression: `hasValue() or (children().count() > id.count())`
    /// 
    /// ## Binding
    /// - **Strength**: required
    /// - **Description**: Type for strand
    /// - **ValueSet**: http://hl7.org/fhir/ValueSet/strand-type|5.0.0
    #[fhir_invariant(key="ele-1", severity="error", human="All FHIR elements must have a @value or children", expr="hasValue() or (children().count() > id.count())", path="MolecularSequence.relative.startingSequence.strand")]
    pub strand: Option<Code>,
}

