// @generated by AtriusValueSetGenerator
// DO NOT EDIT MANUALLY

#![allow(non_camel_case_types)]
#![allow(non_snake_case)]
#![allow(clippy::upper_case_acronyms)]

use super::super::super::{Code, CodeableConcept, Coding, Element, Uri};

use super::super::super::string::String as FhirString;

///FHIR ValueSet: GenomicStudyDataFormat
///Canonical URL: http://hl7.org/fhir/ValueSet/genomicstudy-dataformat
///Version: 5.0.0
///Title: Genomic Study Data Format
///Status: draft
///The data formats relevant to Genomic Study analysis.
///Compose includes 40 explicit concept codes
///Includes systems:
///- http://hl7.org/fhir/genomicstudy-dataformat
pub struct GenomicStudyDataFormat;
impl GenomicStudyDataFormat {
    pub const URL: &'static str = "http://hl7.org/fhir/ValueSet/genomicstudy-dataformat";
    pub const HAS_NONLOCAL_RULES: bool = false;
    pub fn version() -> Option<&'static str> {
        Some("5.0.0")
    }
    pub fn include_systems() -> &'static [&'static str] {
        &["http://hl7.org/fhir/genomicstudy-dataformat"]
    }
    pub fn include_value_sets() -> &'static [&'static str] {
        &[]
    }
    /// Systems that are included as whole CodeSystems but are not locally enumerable.
    ///
    /// If this is non-empty, callers should use a terminology server for definitive validation.
    pub fn include_whole_systems() -> &'static [&'static str] {
        &[]
    }
    /// Returns true only when this ValueSet can be treated as fully locally checkable.
    ///
    /// If `HAS_NONLOCAL_RULES` is true (filters, include.valueSet, or non-enumerated whole-system includes),
    /// local checks are best-effort and may require a terminology server.
    pub fn is_locally_enumerated() -> bool {
        !Self::HAS_NONLOCAL_RULES
            && (!Self::expansion_pairs().is_empty() || !Self::include_pairs().is_empty())
    }
    fn include_pairs() -> &'static [(&'static str, &'static str)] {
        &[
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bam"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bed"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bedpe"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bedgraph"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bigbed"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bigWig"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "birdsuite-files"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "broadpeak"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cbs"),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "chemical-reactivity-probing-profiles",
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "chrom-sizes"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cn"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "custom-file-formats"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cytoband"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "fasta"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gct"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cram"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "genepred"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gff-gtf"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gistic"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "goby"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gwas"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "igv"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "loh"),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "maf-multiple-alignment-format",
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "maf-mutation-annotation-format",
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "merged-bam-file"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "mut"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "narrowpeak"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "psl"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "res"),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "rna-secondary-structure-formats",
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "sam"),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "sample-info-attributes-file",
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "seg"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "tdf"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "track-line"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "type-line"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "vcf"),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "wig"),
        ]
    }
    fn include_entries() -> &'static [(
        &'static str,
        &'static str,
        Option<&'static str>,
        Option<&'static str>,
    )] {
        &[
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bam", Some("BAM"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "bed", Some("BED"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "bedgraph",
                Some("BedGraph"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "bedpe",
                Some("BEDPE"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "bigWig",
                Some("bigWig"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "bigbed",
                Some("bigBed"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "birdsuite-files",
                Some("Birdsuite-Files"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "broadpeak",
                Some("broadPeak"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cbs", Some("CBS"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "chemical-reactivity-probing-profiles",
                Some("Chemical-Reactivity-Probing-Profiles"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "chrom-sizes",
                Some("chrom-sizes"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cn", Some("CN"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "cram", Some("CRAM"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "custom-file-formats",
                Some("Custom-File-Formats"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "cytoband",
                Some("Cytoband"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "fasta",
                Some("FASTA"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gct", Some("GCT"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "genepred",
                Some("genePred"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "gff-gtf",
                Some("GFF/GTF"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "gistic",
                Some("GISTIC"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "goby", Some("Goby"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "gwas", Some("GWAS"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "igv", Some("IGV"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "loh", Some("LOH"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "maf-multiple-alignment-format",
                Some("MAF-Multiple Alignment Format"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "maf-mutation-annotation-format",
                Some("MAF-Mutation-Annotation-Format"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "merged-bam-file",
                Some("Merged BAM File"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "mut", Some("MUT"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "narrowpeak",
                Some("narrowPeak"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "psl", Some("PSL"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "res", Some("RES"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "rna-secondary-structure-formats",
                Some("RNA-Secondary-Structure-Formats"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "sam", Some("SAM"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "sample-info-attributes-file",
                Some("Sample-Info-Attributes-file"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "seg", Some("SEG"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "tdf", Some("TDF"), None),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "track-line",
                Some("Track Line"),
                None,
            ),
            (
                "http://hl7.org/fhir/genomicstudy-dataformat",
                "type-line",
                Some("Type Line"),
                None,
            ),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "vcf", Some("VCF"), None),
            ("http://hl7.org/fhir/genomicstudy-dataformat", "wig", Some("WIG"), None),
        ]
    }
    /// Return the preferred display text for a (system, code) pair, if present in this ValueSet's explicit concepts.
    pub fn display(system: &str, code: &str) -> Option<&'static str> {
        for (s, c, d, _def) in Self::include_entries() {
            if *s == system && *c == code {
                return *d;
            }
        }
        None
    }
    /// Return the ValueSet concept definition for a (system, code) pair, if present.
    ///
    /// In `valuesets.json` this often comes from the extension:
    /// `http://hl7.org/fhir/StructureDefinition/valueset-concept-definition`.
    pub fn definition(system: &str, code: &str) -> Option<&'static str> {
        for (s, c, _d, def) in Self::include_entries() {
            if *s == system && *c == code {
                return *def;
            }
        }
        None
    }
    /// Create a minimal `Coding` from a (system, code) pair.
    ///
    /// If this ValueSet provides an explicit display for the code, it will be populated.
    pub fn to_coding(system: &str, code: &str) -> Coding {
        let sys_uri: Uri = Element {
            id: None,
            extension: None,
            value: Some(system.to_string()),
        };
        let code_el: Code = Element {
            id: None,
            extension: None,
            value: Some(code.to_string()),
        };
        let display = Self::display(system, code)
            .map(|s| FhirString {
                id: None,
                extension: None,
                value: Some(s.to_string()),
            });
        Coding {
            id: None,
            extension: None,
            system: Some(sys_uri),
            version: Self::version()
                .map(|v| FhirString {
                    id: None,
                    extension: None,
                    value: Some(v.to_string()),
                }),
            code: Some(code_el),
            display,
            user_selected: None,
        }
    }
    fn exclude_pairs() -> &'static [(&'static str, &'static str)] {
        &[]
    }
    fn expansion_pairs() -> &'static [(&'static str, &'static str)] {
        &[]
    }
    /// Best-effort local validation for a `Coding` against this ValueSet.
    ///
    /// Rules:
    /// - Explicit excludes (compose.exclude.concept) always deny.
    /// - If expansion.contains exists, it is treated as an explicit allow-list.
    /// - Else if compose.include.concept exists, it is treated as an explicit allow-list.
    /// - Else: fall back to whole-system validation when possible (generated CodeSystem enums / RGB grammar).
    ///
    /// Note: terminology filters and include.valueSet references are NOT evaluated locally.
    pub fn contains_coding(coding: &Coding) -> bool {
        let system = match &coding.system {
            Some(s) => s.value.as_deref().unwrap_or(""),
            None => return false,
        };
        let code = match &coding.code {
            Some(c) => c.value.as_deref().unwrap_or(""),
            None => return false,
        };
        if Self::exclude_pairs().iter().any(|(s, c)| *s == system && *c == code) {
            return false;
        }
        let exp = Self::expansion_pairs();
        if !exp.is_empty() {
            return exp.iter().any(|(s, c)| *s == system && *c == code);
        }
        let inc = Self::include_pairs();
        if !inc.is_empty() {
            return inc.iter().any(|(s, c)| *s == system && *c == code);
        }
        match system {
            "http://hl7.org/fhir/genomicstudy-dataformat" => {
                super::super::code_systems::genomicstudy_dataformat::GenomicStudyDataFormat::try_from_code(
                        code,
                    )
                    .is_ok()
            }
            _ => false,
        }
    }
    /// Best-effort local validation for a `CodeableConcept` against this ValueSet.
    ///
    /// Returns true if any `coding` entry is contained in the ValueSet.
    pub fn contains_codeable_concept(cc: &CodeableConcept) -> bool {
        match cc.coding.as_ref() {
            Some(codings) => codings.iter().any(|c| Self::contains_coding(c)),
            None => false,
        }
    }
    /// Convenience: validate a FHIR `code` primitive when the binding is on `code`.
    ///
    /// This only works reliably when the ValueSet includes exactly one system.
    pub fn contains_code(code: &Code) -> bool {
        let code_str = match code.value.as_deref() {
            Some(v) => v,
            None => return false,
        };
        let systems = Self::include_systems();
        if systems.len() != 1 {
            return false;
        }
        let system: Uri = Element {
            id: None,
            extension: None,
            value: Some(systems[0].to_string()),
        };
        let code: Code = Element {
            id: None,
            extension: None,
            value: Some(code_str.to_string()),
        };
        let coding = Coding {
            id: None,
            extension: None,
            system: Some(system),
            version: None,
            code: Some(code),
            display: None,
            user_selected: None,
        };
        Self::contains_coding(&coding)
    }
    /// Create a minimal `CodeableConcept` (single coding) from a (system, code) pair.
    pub fn to_codeable_concept(system: &str, code: &str) -> CodeableConcept {
        let coding = Self::to_coding(system, code);
        CodeableConcept {
            id: None,
            extension: None,
            coding: Some(vec![coding]),
            text: None,
        }
    }
}
impl super::super::bindings::ValueSetMembership<Code> for GenomicStudyDataFormat {
    fn contains(v: &Code) -> bool {
        Self::contains_code(v)
    }
}
impl super::super::bindings::ValueSetMembership<Coding> for GenomicStudyDataFormat {
    fn contains(v: &Coding) -> bool {
        Self::contains_coding(v)
    }
}
impl super::super::bindings::ValueSetMembership<CodeableConcept>
for GenomicStudyDataFormat {
    fn contains(v: &CodeableConcept) -> bool {
        Self::contains_codeable_concept(v)
    }
}
fn is_rgb_hex(code: &str) -> bool {
    let b = code.as_bytes();
    if b.len() != 7 || b[0] != b'#' {
        return false;
    }
    fn is_hex(x: u8) -> bool {
        (b'0'..=b'9').contains(&x) || (b'a'..=b'f').contains(&x)
            || (b'A'..=b'F').contains(&x)
    }
    is_hex(b[1]) && is_hex(b[2]) && is_hex(b[3]) && is_hex(b[4]) && is_hex(b[5])
        && is_hex(b[6])
}

