// @generated by AtriusValueSetGen
// DO NOT EDIT MANUALLY

#![allow(non_camel_case_types)]
#![allow(non_snake_case)]
#![allow(clippy::upper_case_acronyms)]

use serde::{Deserialize, Serialize};

use super::super::super::{Boolean, Code, CodeableConcept, Coding, Element, Uri};

use super::super::super::string::String as FhirString;

///FHIR CodeSystem: GenomicStudyDataFormat
///Canonical URL: http://hl7.org/fhir/genomicstudy-dataformat
///Version: 5.0.0
///Title: Genomic Study Data Format
///Status: active
///The data format relevant to genomics. These formats and relevant codes were pulled from [Integrative Genomics Viewer Documentation](https://software.broadinstitute.org/software/igv/FileFormats) by Broad Institute.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum GenomicStudyDataFormat {
    ///Display: BAM
    #[serde(rename = "bam")]
    Bam,
    ///Display: BED
    #[serde(rename = "bed")]
    Bed,
    ///Display: BEDPE
    #[serde(rename = "bedpe")]
    Bedpe,
    ///Display: BedGraph
    #[serde(rename = "bedgraph")]
    Bedgraph,
    ///Display: bigBed
    #[serde(rename = "bigbed")]
    Bigbed,
    ///Display: bigWig
    #[serde(rename = "bigWig")]
    Bigwig,
    ///Display: Birdsuite-Files
    #[serde(rename = "birdsuite-files")]
    BirdsuiteFiles,
    ///Display: broadPeak
    #[serde(rename = "broadpeak")]
    Broadpeak,
    ///Display: CBS
    #[serde(rename = "cbs")]
    Cbs,
    ///Display: Chemical-Reactivity-Probing-Profiles
    #[serde(rename = "chemical-reactivity-probing-profiles")]
    ChemicalReactivityProbingProfiles,
    ///Display: chrom-sizes
    #[serde(rename = "chrom-sizes")]
    ChromSizes,
    ///Display: CN
    #[serde(rename = "cn")]
    Cn,
    ///Display: Custom-File-Formats
    #[serde(rename = "custom-file-formats")]
    CustomFileFormats,
    ///Display: Cytoband
    #[serde(rename = "cytoband")]
    Cytoband,
    ///Display: FASTA
    #[serde(rename = "fasta")]
    Fasta,
    ///Display: GCT
    #[serde(rename = "gct")]
    Gct,
    ///Display: CRAM
    #[serde(rename = "cram")]
    Cram,
    ///Display: genePred
    #[serde(rename = "genepred")]
    Genepred,
    ///Display: GFF/GTF
    #[serde(rename = "gff-gtf")]
    GffGtf,
    ///Display: GISTIC
    #[serde(rename = "gistic")]
    Gistic,
    ///Display: Goby
    #[serde(rename = "goby")]
    Goby,
    ///Display: GWAS
    #[serde(rename = "gwas")]
    Gwas,
    ///Display: IGV
    #[serde(rename = "igv")]
    Igv,
    ///Display: LOH
    #[serde(rename = "loh")]
    Loh,
    ///Display: MAF-Multiple Alignment Format
    #[serde(rename = "maf-multiple-alignment-format")]
    MafMultipleAlignmentFormat,
    ///Display: MAF-Mutation-Annotation-Format
    #[serde(rename = "maf-mutation-annotation-format")]
    MafMutationAnnotationFormat,
    ///Display: Merged BAM File
    #[serde(rename = "merged-bam-file")]
    MergedBamFile,
    ///Display: MUT
    #[serde(rename = "mut")]
    Mut,
    ///Display: narrowPeak
    #[serde(rename = "narrowpeak")]
    Narrowpeak,
    ///Display: PSL
    #[serde(rename = "psl")]
    Psl,
    ///Display: RES
    #[serde(rename = "res")]
    Res,
    ///Display: RNA-Secondary-Structure-Formats
    #[serde(rename = "rna-secondary-structure-formats")]
    RnaSecondaryStructureFormats,
    ///Display: SAM
    #[serde(rename = "sam")]
    Sam,
    ///Display: Sample-Info-Attributes-file
    #[serde(rename = "sample-info-attributes-file")]
    SampleInfoAttributesFile,
    ///Display: SEG
    #[serde(rename = "seg")]
    Seg,
    ///Display: TDF
    #[serde(rename = "tdf")]
    Tdf,
    ///Display: Track Line
    #[serde(rename = "track-line")]
    TrackLine,
    ///Display: Type Line
    #[serde(rename = "type-line")]
    TypeLine,
    ///Display: VCF
    #[serde(rename = "vcf")]
    Vcf,
    ///Display: WIG
    #[serde(rename = "wig")]
    Wig,
}
impl GenomicStudyDataFormat {
    pub fn as_code(&self) -> &'static str {
        match self {
            GenomicStudyDataFormat::Bam => "bam",
            GenomicStudyDataFormat::Bed => "bed",
            GenomicStudyDataFormat::Bedpe => "bedpe",
            GenomicStudyDataFormat::Bedgraph => "bedgraph",
            GenomicStudyDataFormat::Bigbed => "bigbed",
            GenomicStudyDataFormat::Bigwig => "bigWig",
            GenomicStudyDataFormat::BirdsuiteFiles => "birdsuite-files",
            GenomicStudyDataFormat::Broadpeak => "broadpeak",
            GenomicStudyDataFormat::Cbs => "cbs",
            GenomicStudyDataFormat::ChemicalReactivityProbingProfiles => {
                "chemical-reactivity-probing-profiles"
            }
            GenomicStudyDataFormat::ChromSizes => "chrom-sizes",
            GenomicStudyDataFormat::Cn => "cn",
            GenomicStudyDataFormat::CustomFileFormats => "custom-file-formats",
            GenomicStudyDataFormat::Cytoband => "cytoband",
            GenomicStudyDataFormat::Fasta => "fasta",
            GenomicStudyDataFormat::Gct => "gct",
            GenomicStudyDataFormat::Cram => "cram",
            GenomicStudyDataFormat::Genepred => "genepred",
            GenomicStudyDataFormat::GffGtf => "gff-gtf",
            GenomicStudyDataFormat::Gistic => "gistic",
            GenomicStudyDataFormat::Goby => "goby",
            GenomicStudyDataFormat::Gwas => "gwas",
            GenomicStudyDataFormat::Igv => "igv",
            GenomicStudyDataFormat::Loh => "loh",
            GenomicStudyDataFormat::MafMultipleAlignmentFormat => {
                "maf-multiple-alignment-format"
            }
            GenomicStudyDataFormat::MafMutationAnnotationFormat => {
                "maf-mutation-annotation-format"
            }
            GenomicStudyDataFormat::MergedBamFile => "merged-bam-file",
            GenomicStudyDataFormat::Mut => "mut",
            GenomicStudyDataFormat::Narrowpeak => "narrowpeak",
            GenomicStudyDataFormat::Psl => "psl",
            GenomicStudyDataFormat::Res => "res",
            GenomicStudyDataFormat::RnaSecondaryStructureFormats => {
                "rna-secondary-structure-formats"
            }
            GenomicStudyDataFormat::Sam => "sam",
            GenomicStudyDataFormat::SampleInfoAttributesFile => {
                "sample-info-attributes-file"
            }
            GenomicStudyDataFormat::Seg => "seg",
            GenomicStudyDataFormat::Tdf => "tdf",
            GenomicStudyDataFormat::TrackLine => "track-line",
            GenomicStudyDataFormat::TypeLine => "type-line",
            GenomicStudyDataFormat::Vcf => "vcf",
            GenomicStudyDataFormat::Wig => "wig",
        }
    }
    pub fn system() -> &'static str {
        "http://hl7.org/fhir/genomicstudy-dataformat"
    }
    pub fn version() -> Option<&'static str> {
        Some("5.0.0")
    }
    /// Convert this code into a FHIR `code` primitive (`Code = Element<String, Extension>`).
    ///
    /// Useful for elements like `Patient.gender` that are bound directly to a ValueSet and use
    /// the `code` datatype.
    pub fn to_code(self) -> Code {
        Element {
            id: None,
            extension: None,
            value: Some(self.as_code().to_string()),
        }
    }
    /// Convert this code into a FHIR `Coding` with `system` and (if available) `version` set.
    ///
    /// Useful for `Coding` fields and for inclusion inside `CodeableConcept.coding`.
    pub fn to_coding(self) -> Coding {
        let system: Uri = Element {
            id: None,
            extension: None,
            value: Some(Self::system().to_string()),
        };
        Coding {
            id: None,
            extension: None,
            system: Some(system),
            version: Self::version()
                .map(|v| FhirString {
                    id: None,
                    extension: None,
                    value: Some(v.to_string()),
                }),
            code: Some(self.to_code()),
            display: None,
            user_selected: Option::<Boolean>::None,
        }
    }
    /// Convert this code into a minimal `CodeableConcept` (with a single `coding`).
    ///
    /// This is especially useful for bound elements that use `CodeableConcept` (e.g.
    /// extensible bindings).
    pub fn to_codeable_concept(self) -> CodeableConcept {
        CodeableConcept {
            id: None,
            extension: None,
            coding: Some(vec![self.to_coding()]),
            text: None,
        }
    }
    /// Parse a code string into this enum.
    ///
    /// This is a convenience wrapper around the `TryFrom<&str>` implementation.
    pub fn try_from_code(code: &str) -> Result<Self, ()> {
        <Self as core::convert::TryFrom<&str>>::try_from(code)
    }
    /// Parse from a FHIR `code` primitive (`Code = Element<String, Extension>`).
    pub fn try_from_code_element(code: &Code) -> Result<Self, ()> {
        match code.value.as_deref() {
            Some(v) => Self::try_from_code(v),
            None => Err(()),
        }
    }
}
impl core::convert::TryFrom<&str> for GenomicStudyDataFormat {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, <Self as core::convert::TryFrom<&str>>::Error> {
        match s {
            "bam" => Ok(GenomicStudyDataFormat::Bam),
            "bed" => Ok(GenomicStudyDataFormat::Bed),
            "bedpe" => Ok(GenomicStudyDataFormat::Bedpe),
            "bedgraph" => Ok(GenomicStudyDataFormat::Bedgraph),
            "bigbed" => Ok(GenomicStudyDataFormat::Bigbed),
            "bigWig" => Ok(GenomicStudyDataFormat::Bigwig),
            "birdsuite-files" => Ok(GenomicStudyDataFormat::BirdsuiteFiles),
            "broadpeak" => Ok(GenomicStudyDataFormat::Broadpeak),
            "cbs" => Ok(GenomicStudyDataFormat::Cbs),
            "chemical-reactivity-probing-profiles" => {
                Ok(GenomicStudyDataFormat::ChemicalReactivityProbingProfiles)
            }
            "chrom-sizes" => Ok(GenomicStudyDataFormat::ChromSizes),
            "cn" => Ok(GenomicStudyDataFormat::Cn),
            "custom-file-formats" => Ok(GenomicStudyDataFormat::CustomFileFormats),
            "cytoband" => Ok(GenomicStudyDataFormat::Cytoband),
            "fasta" => Ok(GenomicStudyDataFormat::Fasta),
            "gct" => Ok(GenomicStudyDataFormat::Gct),
            "cram" => Ok(GenomicStudyDataFormat::Cram),
            "genepred" => Ok(GenomicStudyDataFormat::Genepred),
            "gff-gtf" => Ok(GenomicStudyDataFormat::GffGtf),
            "gistic" => Ok(GenomicStudyDataFormat::Gistic),
            "goby" => Ok(GenomicStudyDataFormat::Goby),
            "gwas" => Ok(GenomicStudyDataFormat::Gwas),
            "igv" => Ok(GenomicStudyDataFormat::Igv),
            "loh" => Ok(GenomicStudyDataFormat::Loh),
            "maf-multiple-alignment-format" => {
                Ok(GenomicStudyDataFormat::MafMultipleAlignmentFormat)
            }
            "maf-mutation-annotation-format" => {
                Ok(GenomicStudyDataFormat::MafMutationAnnotationFormat)
            }
            "merged-bam-file" => Ok(GenomicStudyDataFormat::MergedBamFile),
            "mut" => Ok(GenomicStudyDataFormat::Mut),
            "narrowpeak" => Ok(GenomicStudyDataFormat::Narrowpeak),
            "psl" => Ok(GenomicStudyDataFormat::Psl),
            "res" => Ok(GenomicStudyDataFormat::Res),
            "rna-secondary-structure-formats" => {
                Ok(GenomicStudyDataFormat::RnaSecondaryStructureFormats)
            }
            "sam" => Ok(GenomicStudyDataFormat::Sam),
            "sample-info-attributes-file" => {
                Ok(GenomicStudyDataFormat::SampleInfoAttributesFile)
            }
            "seg" => Ok(GenomicStudyDataFormat::Seg),
            "tdf" => Ok(GenomicStudyDataFormat::Tdf),
            "track-line" => Ok(GenomicStudyDataFormat::TrackLine),
            "type-line" => Ok(GenomicStudyDataFormat::TypeLine),
            "vcf" => Ok(GenomicStudyDataFormat::Vcf),
            "wig" => Ok(GenomicStudyDataFormat::Wig),
            _ => Err(()),
        }
    }
}
impl core::fmt::Display for GenomicStudyDataFormat {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.write_str(self.as_code())
    }
}

